text/0000-layer-metadata-files.md

```diff
@@ -0,0 +1,184 @@
+# Meta
+[meta]: #meta
+- Name: Combine and organize metadata file locations across lifecycle
+- Start Date: 2021-03-11
+- Author(s): [jabrown85](https://github.com/jabrown85) [samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This is a proposal to change the default locations of several files and directories. Part of the reasoning behind this is to disambiguate metadata files from application or layer files and avoid naming collisions.
+
+# Motivation
+[motivation]: #motivation
+
+Currently all the metadata files live under the `<layers>` folder. This means that application metadata files like `launch.toml`, `build.toml` or `store.toml` share the same directory as layer metadata files. Since the layers could possibly be named `launch`, `build` or `store` this leads to ambiguity, conflicts and/or restrictions on layer names and their associated metadata files that should not exist. This also prevents us from possibly adding new metadata files in the future without breaking backwards compatibility with buildpacks who may have been creating layers with the same name. This proposal seeks to disambiguate special metadata files from free-form layer metadata files.
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is as follows -
+
+The current folder structure is -
+
+```
+/cnb (provided by builder image)
+├── buildpacks
+│   └── <buildpacks>
+├── lifecycle
+│   └── <creator/builder/etc>
+├── order.toml
+└── stack.toml
+/layers (configurable with `CNB_LAYERS_DIR` or -layers)
+├── group.toml
+├── plan.toml
+├── order.toml (written by a platform, optionally)
+├── report.toml
+├── <escaped_buildpack_id> (buildpack arg `$1`)
+│   ├── build.toml
+│   ├── launch.toml
+│   ├── store.toml
+│   ├── <layer>.toml
+│   └── <layer>
+└── config
+    └── metadata.toml
+/workspace (configurable with `CNB_APP_DIR ` or -app)
+└── <app source code to transform>
+```
+
+This proposal is to change this folder structure to -
+
+```
+/cnb (provided by builder image)
+├── buildpacks
+│   └── <buildpacks>
+├── lifecycle
+│   └── <creator/builder/etc>
+├── order.toml
+└── stack.toml
+/workspace (configurable with `CNB_WORKSPACE_DIR`)
+├── app (configurable with `CNB_APP_DIR ` or -app)
+│    └── <app source code to transform>
+├── platform-config (configurable with `CNB_PLATFORM_CONFIG_DIR`)
+│   └── order.toml (written by a platform, optionally - prior to `lifecycle` execution `platform/prepare`)
+├── lifecycle-config (configurable with `CNB_LIFECYCLE_CONFIG_DIR`)
+│   └── group.toml
+│   └── plan.toml
+│   └── report.toml
+├── layers (configurable with `CNB_LAYERS_DIR` or -layers)
+│   └── new-buildpack (`CNB_BP_LAYERS_DIR`)
```

Comment:
Since `CNB_BP_LAYERS_DIR` is used more often than `CNB_LAYERS_DIR`, could we pick a shorter name for the former?

Maybe something like:
- `CNB_ALL_LAYERS_DIR` for `/workspace/layers`
- `CNB_ALL_CONFIG_DIR` for `/workspace/layers-config`
- `CNB_LAYERS_DIR` for `/workspace/layers/<buildpack-id>`
- `CNB_CONFIG_DIR` for `/workspace/layers-config/<buildpack-id>`?

Reply:
I have a minor concern with using `CNB_LAYERS_DIR` for `CNB_BP_LAYERS_DIR` since the former is already a config var for `/layers`. We would be moving `CNB_LAYERS_DIR` from a platform env var to a buildpack env var, which feels a bit weird.

Reply:
Ah, good point. Maybe `CNB_ALL_LAYER_DIR` and `CNB_LAYER_DIR`?

Reply:
I still like `CNB_BP_LAYERS_DIR` for its obviousness.

Reply:
Right now the buildpack's layers dir is provided to it as a positional arg

> /bin/build <layers[EIC]> <platform[AR]> <plan[ER]>, Working Dir: <app[AI]>

Are we proposing replacing all positional args in the buildpack API with environment variables as part of that RFC? J I think that is a good idea but it's a big enough change that I think we should call this out explicitly in a separate section and provide the names for the env vars that will replace the plan and platform arguments as well.

Reply:
If we go with `CNB_BP_LAYERS_DIR` should we switch from `CNB_BUILDPACK_DIR` to `CNB_BP_DIR` for consistency?

Reply:
I decided to change them to `_BUILDPACK_` to match `CNB_BUILDPACK_DIR`. They are long, but explicit and match our current standards that buildpacks are already built against.

Reply:
@ekcasey I would prefer to not tackle the positional argument changes in this RFC.

Reply:
If we aren't switching from position arguments to environment variables then I am unsure what `CNB_BUILDPACK_LAYERS_DIR` is referring to above. The layers dir is currently a positional argument and nobody would be setting or reading this environment variable.

Reply:
Also, if we are introducing at least one environment variable `CNB_BUILDPACK_CONFIG_DIR` it does seem to me like an appropriate time to convert the rest (just for the sake of internal consistency).